### PR TITLE
Ellipses

### PR DESCRIPTION
**Pull request by JamesCraster**
This PR adds matching of source and target vertices to the networkx graph, uses the elliptical bounding box method to limit the graph size, and applies dijkstra's algorithm to produce the shortest path between the source and target.

## Type of change

- [ ] Bug fix
- [x] New feature
- [ ] Documentation
- [x] New algorithm

## Tests
How has the code been tested?

## Issues fixes

List any issues this closes:

## Checklist
Please ensure you have done the following:

- [ ] Code conforms to the projects style guidelines

- [ ] New code documented using [google docstyle](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html)

- [ ] Updated documentation of any changed code

- [ ] Any additional dependencies have been added to the appropriate location (e.g. `setup.py` or `requirements.txt`)

- [ ] Unit tests written and passing. Tests should prove that fix is effective or new feature works as expected

- [ ] Code needs to run on the cluster? Tag database administrator in the pull request review to configure.

### CODE REVIEW

File: entrypoints/air_pollution_shortest_path.py [hunk 1]
```diff
@@ -1,13 +1,106 @@
...
+from urbanroute.geospatial import update_cost, ellipse_bounding_box
+from urbanroute.queries import HexGridQuery
+
+
+def main(  # pylint: disable=too-many-arguments
+    secretfile: str,
+    instance_id: str = "d5e691ef9a1f2e86743f614806319d93e30709fe179dfb27e7b99b9b967c8737",
+    source: Tuple[float, float] = (51.4929, -0.1215),
+    start_time: Optional[str] = "2020-01-24T09:00:00",
+    target: Tuple[float, float] = (51.4929, -0.2215),
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:28]: Change to 2 separtate arguments

File: entrypoints/air_pollution_shortest_path.py [hunk 2]
```diff
@@ -1,13 +1,106 @@
...
+    upto_time: Optional[str] = "2020-01-24T10:00:00",
+    verbose: Optional[bool] = False,
+):
+    """
+    instance_id: Id of the air quality trained model.
+
+    secretfile: Path to the database secretfile.
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:35]: Add to docs

File: entrypoints/air_pollution_shortest_path.py [hunk 3]
```diff
@@ -1,13 +1,106 @@
...
+        G: nx.MultiDiGraph = ox.graph_from_bbox(box[0], box[1], box[2], box[3])
+        G = G.to_directed()
+
+        # snap source and target to the graph
+        newSource = ox.distance.get_nearest_node(G, source)
+        newTarget = ox.distance.get_nearest_node(G, target)
+
+        # fig, ax = ox.plot_graph(G, show=False, close=False)
+        # ax.scatter(source[1], source[0], c='red')
+        # ax.scatter(target[1], target[0], c='red')
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:73]: Lets move visualisations to a new `notebooks` directory. 

Directory should contain a requirements.txt / `.yml` with depedencies such as jupyter & matplotlib

File: entrypoints/air_pollution_shortest_path.py [hunk 4]
```diff
@@ -1,13 +1,106 @@
...
+    gdf = gdf.rename(columns=dict(geom="geometry"))
+    gdf.crs = "EPSG:4326"
+    # gdf = gpd.GeoDataFrame(result_df, crs=4326, geometry="geom")
+    logger.info("%s rows in hexgrid results", len(gdf))
+
+    if source is not None and target is not None:
+        # use bounding box of surrounding ellipse to limit graph size
+        box = ellipse_bounding_box((source[1], source[0]), (target[1], target[0]))
+        G: nx.MultiDiGraph = ox.graph_from_bbox(box[0], box[1], box[2], box[3])
+        G = G.to_directed()
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:65]: Remove for now. Create a new issue for implementing A*:

- need to change Euclidean distance for haversine.
- how would this work for pollution?
- how does this work for multi di graphs?

File: entrypoints/air_pollution_shortest_path.py [hunk 5]
```diff
@@ -1,13 +1,106 @@
...
+            for edge in G.edges
+        ]
+        norm = colors.Normalize(vmin=min(ev), vmax=max(ev))
+        cmap = cm.ScalarMappable(norm=norm, cmap=cm.inferno)
+        ec = [cmap.to_rgba(cl) for cl in ev]
+        fig, ax = ox.plot_graph_route(
+            G,
+            nx.dijkstra_path(G, newSource, newTarget, weight="NO2_mean"),
+            edge_color=ec,
+        )
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:100]: Move the viz into a notebook.

File: urbanroute/urbanroute/geospatial/ellipses.py [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+from shapely.geometry import Point, Polygon
+import shapely.affinity
+from descartes import PolygonPatch
+from typing import Tuple, List
+import osmnx as ox
+
+
+def ellipse_bounding_box(
+    source: Tuple[float, float], dest: Tuple[float, float], tau=1.1
+) -> List[Tuple[float, float]]:
```
Comment (PatrickOHara) [inline urbanroute/urbanroute/geospatial/ellipses.py:12]: tau: Optional[float] = 1.1

Comment (PatrickOHara) [inline urbanroute/urbanroute/geospatial/ellipses.py:12]: Change return type of bounding box to [N,S,E,W]

File: urbanroute/urbanroute/geospatial/ellipses.py [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+
+        Args:
+            source, target: two points in longitude, latitude format
+            tau: a constant to be multiplied by the distance between the source and target; the optimal path
+                 is assumed to be smaller than this product.
+        
+        Returns: 
+            A bounding box that surrounds the source and target, in north, south, east, west format,
+            includes all paths of length at most: tau multiplied by the distance between the points.
+            This inclusion is done by creating an implicit elipse that the source and dest are foci of.
```
Comment (PatrickOHara) [inline urbanroute/urbanroute/geospatial/ellipses.py:24]: No indentation- move left one tab

Comment (PatrickOHara) [reply urbanroute/urbanroute/geospatial/ellipses.py:24]: Add link to paper in the function docs

File: urbanroute/urbanroute/geospatial/ellipses.py [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+    """
+    theta = math.atan((dest[1] - source[1]) / (dest[0] - source[0]))
+    tau = 1.1
+    distance = ox.distance.great_circle_vec(source[1], source[0], dest[1], dest[0])
+    b = (dest[1] + source[1]) / 2
+    a = (dest[0] + source[0]) / 2
+    A = (tau / 2) * math.sqrt((dest[1] - source[1]) ** 2 + (dest[0] - source[0]) ** 2)
+    B = math.sqrt(
+        A ** 2 - ((dest[1] - source[1]) ** 2 + (dest[0] - source[0]) ** 2) / 4
+    )
```
Comment (PatrickOHara) [inline urbanroute/urbanroute/geospatial/ellipses.py:34]: Comment high level algorithm

File: urbanroute/urbanroute/geospatial/ellipses.py [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+        Returns: 
+            A bounding box that surrounds the source and target, in north, south, east, west format,
+            includes all paths of length at most: tau multiplied by the distance between the points.
+            This inclusion is done by creating an implicit elipse that the source and dest are foci of.
+    """
+    theta = math.atan((dest[1] - source[1]) / (dest[0] - source[0]))
+    tau = 1.1
+    distance = ox.distance.great_circle_vec(source[1], source[0], dest[1], dest[0])
+    b = (dest[1] + source[1]) / 2
+    a = (dest[0] + source[0]) / 2
```
Comment (PatrickOHara) [inline urbanroute/urbanroute/geospatial/ellipses.py:30]: Better naming for variables

File: urbanroute/urbanroute/geospatial/ellipses.py (hunk 8 above)
Comment (PatrickOHara) [reply urbanroute/urbanroute/geospatial/ellipses.py:34]: Rename A, B to axis

**Comment by PatrickOHara** [comment]
If source_lat == target_lat will this break due to divide by 0?

Add a quick test.

**Comment by JamesCraster** [comment]
> If source_lat == target_lat will this break due to divide by 0?
> 
> Add a quick test.

This occurs if source_long = target_long, in which case it should be that theta = pi/2

File: entrypoints/air_pollution_shortest_path.py [hunk 10]
```diff
@@ -1,39 +1,55 @@
...
 
-from typing import Optional
+from typing import Optional, Tuple
 import logging
 import typer
 import geopandas as gpd
 import osmnx as ox
 import networkx as nx
+import math
+from cleanair.databases.queries import AirQualityResultQuery
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:10]: Looks like the merge went wrong. I think I removed this

File: entrypoints/air_pollution_shortest_path.py [hunk 11]
```diff
@@ -57,16 +73,96 @@ def main(  # pylint: disable=too-many-arguments
...
+
+        logger.info("Mapping air quality predictions to the road network.")
+        G = update_cost(G, gdf, cost_attr="NO2_mean", weight_attr="length")
+        logger.debug("Printing basic stats for the graph:")
+        logger.debug(ox.stats.basic_stats(G))
+
+        for i, (u, v, k, data) in enumerate(G.edges(keys=True, data=True)):
+            if i > 10:
+                break
+            print(u, v, k, data)
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:97]: This should also have been removed in the merge I think?

File: entrypoints/air_pollution_shortest_path.py [hunk 12]
```diff
@@ -57,16 +73,96 @@ def main(  # pylint: disable=too-many-arguments
...
+        
+        route = nx.dijkstra_path(G, newSource, newTarget, weight="NO2_mean")
+        print(route)
 
 if __name__ == "__main__":
     typer.run(main)
+
+@app.get("/route/")
+async def get_route(source_lat:float, source_long:float,target_lat:float,target_long:float):
```
Comment (PatrickOHara) [inline entrypoints/air_pollution_shortest_path.py:106]: Is pretty much a direct copy from the above main function? If so lets create a new function we re-uses the repeated functionality?

File: urbanroute/urbanroute/geospatial/ellipses.py [hunk 13]
```diff
@@ -0,0 +1,46 @@
+import matplotlib.pyplot as plt
+import math
+from shapely.geometry import Point, Polygon
+import shapely.affinity
```
Comment (PatrickOHara) [inline urbanroute/urbanroute/geospatial/ellipses.py:4]: Is shapely in the urbanroute setup.py file? Or is it already included due to geopandas & osmnx?

**Pull request merged**